### PR TITLE
[ckpt] fix: normalize PEFT keys in HF save and dump LoRA adapter with live config

### LINKED ISSUE #4761: [Bug] no merged model under the folder `huggingface`
### System Info

------------Pip Info-----------
Version      : 25.3
vllm         : 0.11.0
sglang       : 0.5.2
ray          : 2.53.0
torch        : 2.8.0
----------verl Info-----------
Version      : 0.7.0.dev
Commit Hash  : 9037d4aba4901390923e96d8cd853169c6147dee
----------Platform Info----------
Platform     : Linux-6.8.0-55-generic-x86_64-with-glibc2.35
system       : Linux

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

When using `actor_rollout_ref.actor.checkpoint.save_contents=['model','optimizer','extra','hf_model']`, the model weights in the `<checkpoint folder>/global_step_*/actor/huggingface` folder is not the fully-merged model.

The `model.safetensors.index.json` is like:
```
"weight_map": {
    "base_model.model.lm_head.weight": "model-00007-of-00007.safetensors",
    "base_model.model.model.embed_tokens.weight": "model-00001-of-00007.safetensors",
    "base_model.model.model.layers.0.input_layernorm.weight": "model-00002-of-00007.safetensors",
    "base_model.model.model.layers.0.mlp.down_proj.base_layer.weight": "model-00002-of-00007.safetensors",
    "base_model.model.model.layers.0.mlp.down_proj.lora_A.default.weight": "model-00002-of-00007.safetensors",
    "base_model.model.model.layers.0.mlp.down_proj.lora_B.default.weight": "model-00002-of-00007.safetensors",
...
```

Note:
1. extra prefix `base_model` compared to the base model
2. there are 'base_layer.weight', 'lora_A.default', and 'lora_B.default' for each trainable layer

Command:
```bash
python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_files=$dataset_train \
    data.val_files=$dataset_val \
    data.train_batch_size=192 \
    data.max_prompt_length=2048 \
    data.max_response_length=2048 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    data.image_key=images \
    actor_rollout_ref.model.path=$base_model_hf \
    actor_rollout_ref.actor.optim.lr=3e-6 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=96 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.model.lora_rank=32 \
    actor_rollout_ref.model.lora_alpha=16 \
    actor_rollout_ref.model.target_modules=all-linear \
    actor_rollout_ref.model.exclude_modules='.*visual.*' \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.01 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.actor.checkpoint.save_contents=['model','optimizer','extra','hf_model'] \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=2 \
    actor_rollout_ref.rollout.name=$ENGINE \
    +actor_rollout_ref.rollout.engine_kwargs.vllm.disable_mm_preprocessor_cache=True \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
    actor_rollout_ref.rollout.enable_chunked_prefill=False \
    actor_rollout_ref.rollout.enforce_eager=False \
    actor_rollout_ref.rollout.free_cache_engine=False \
    actor_rollout_ref.rollout.n=8 \
    actor_rollout_ref.rollout.load_format="safetensors" \
    actor_rollout_ref.rollout.layered_summon=True \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=4 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    algorithm.use_kl_in_reward=False \
    trainer.critic_warmup=0 \
    trainer.logger='["console","wandb"]' \
    trainer.project_name=$wandb_project \
    trainer.experiment_name=$exp_name \
    trainer.n_gpus_per_node=6 \
    trainer.nnodes=1 \
    trainer.save_freq=20 \
    trainer.test_freq=5 \
    trainer.total_epochs=1000 \
    trainer.max_actor_ckpt_to_keep=10 \
    trainer.default_local_dir=$default_local_dir \
    trainer.validation_data_dir=$validation_data_dir \
    custom_reward_function.path=$reward_function_path \
    custom_reward_function.name=$reward_function_name \
    reward_model.use_reward_loop=True \
    data.custom_cls.path=$custom_cls_path \
    data.custom_cls.name=$custom_cls_name \
    $@
```

### Expected behavior

The current saved model can not be used directly by vLLM or maybe HF.

Base model layer index is like:
```
 "weight_map": {
    "lm_head.weight": "model-00005-of-00005.safetensors",
    "model.embed_tokens.weight": "model-00001-of-00005.safetensors",
    "model.layers.0.input_layernorm.weight": "model-00001-of-00005.safetensors",
    "model.layers.0.mlp.down_proj.weight": "model-00001-of-00005.safetensors",
    "model.layers.0.mlp.gate_proj.weight": "model-00001-of-00005.safetensors",
    "model.layers.0.mlp.up_proj.weight": "model-00001-of-00005.safetensors",
    "model.layers.0.post_attention_layernorm.weight": "model-00001-of-00005.safetensors",
    "model.layers.0.self_attn.k_proj.bias": "model-00001-of-00005.safetensors",
    "model.layers.0.self_attn.k_proj.weight": "model-00001-of-00005.safetensors",
    "model.layers.0.self_attn.o_proj.weight": "model-00001-of-00005.safetensors",
    "model.layers.0.self_attn.q_proj.bias": "model-00001-of-00005.safetensors",
    "model.layers.0.self_attn.q_proj.weight": "model-00001-of-00005.safetensors",
    "model.layers.0.self_attn.v_proj.bias": "model-00001-of-00005.safetensors",
```

### PR DESCRIPTION
**Pull request by yurekami**
### What does this PR do?

Makes the `huggingface/` output of a LoRA/PEFT training run directly loadable, and dumps the LoRA adapter with its real config.

On main, `FSDPCheckpointManager.save_checkpoint` saves the HF model via `get_fsdp_full_state_dict(...)` → `save_pretrained(hf_local_path, state_dict=state_dict)` with no PEFT handling, so for LoRA runs `huggingface/` contains `base_model.model.*` / `*.base_layer.weight` keys plus raw `lora_A/lora_B` tensors — HF/vLLM cannot load it (see @thvasilo's docker repro in the comments). Since the engine refactor removed `fsdp_workers.py`, main also no longer writes any `lora_adapter/` at save time.

This PR, in the `should_save_hf_model` path, when the model has a `peft_config`:

1. writes a single `{local_path}/lora_adapter/` via `peft.get_peft_model_state_dict` + `peft_config.save_pretrained` — so `adapter_config.json` carries the **live** `lora_alpha`/`task_type` (the original version of this PR reconstructed the config from tensor shapes, which produced `lora_alpha=0`; that is gone)
2. normalizes the state dict with the existing `verl.utils.fsdp_utils.normalize_peft_param_name` (already used by rollout weight-sync) before `save_pretrained`, so `huggingface/` holds plain base-model keys

Note for reviewers: **this is effectively a new diff under the old PR number.** Every code site the original January diff touched was removed in the engine refactor; the branch was rebuilt on current main rather than rebased.

### Test

`tests/utils/ckpt/test_lora_hf_save_on_cpu.py` (CPU, runs in `cpu_unit_tests`):
- normalized full state dict of a real `PeftModel` loads into the plain base model with no missing/unexpected keys
- `_save_lora_adapter` writes `adapter_config.json` with the live `lora_alpha` (32, not 0) and the adapter round-trips through `PeftModel.from_pretrained` with identical weights
- no-op (and no directory) for non-PEFT models

Local testing is CPU-only; multi-GPU FSDP validation is left to reviewers/CI with hardware.

### Checklist

- [x] New CPU unit test added
- [x] `ruff check` / `ruff format` clean (pinned v0.12.2)
- [x] AI-assisted; the submitting human has reviewed every changed line

### CODE REVIEW

File: tests/utils/test_lora_checkpoint_on_cpu.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        for name in list(state_dict.keys()):
+            key = (
+                name.replace("base_model.model.", "")
+                .replace(".base_layer.weight", ".weight")
+                .replace(".base_layer.bias", ".bias")
+            )
+            if key != name:
+                state_dict[key] = state_dict.pop(name)
+
+        return lora_path
```
[the remark this answers is not in the document]
Comment (vermouth1992) [reply tests/utils/test_lora_checkpoint_on_cpu.py]: @yurekami Could you resolve the concerns raised by Gemini?

**Comment by yurekami** [comment]
Thanks for the review feedback! I've addressed both concerns in commit 3952342b:

### 1. Unsafe String Replacement (Fixed)

Replaced the chained `str.replace()` calls with explicit `startswith()`/`endswith()` checks:

```python
# Before (unsafe - replaces ALL occurrences)
key = name.replace("base_model.model.", "").replace(".base_layer.weight", ".weight")

# After (safe - only handles prefix/suffix)
if key.startswith("base_model.model."):
    key = key[len("base_model.model."):]
if key.endswith(".base_layer.weight"):
    key = key[:-len(".base_layer.weight")] + ".weight"
```

This prevents unintended substitutions if patterns appear in the middle of key names.

### 2. Code Duplication (Fixed)

- Converted `_save_lora_adapter()` to a `@staticmethod`
- Removed the duplicated `_save_lora_adapter_standalone()` from the test file (50 lines deleted)
- Tests now call `FSDPCheckpointManager._save_lora_adapter()` directly

All tests pass ✅

**Comment by thvasilo** [comment]
Hi @vermouth1992 could you take another look, this is currently a blocker for us

Also to clarify loading just the adapter with the _original_ base model works, but trying to load the contents of the `huggingface` folder leads to this error

this causes the error

```
docker run --gpus all   -v ~/.cache/huggingface:/root/.cache/huggingface   \
  -v /my-model-output/global_step_4/actor/:/model   \
  --env "HF_TOKEN=$HF_TOKEN"   -p 8000:8000   --ipc=host   vllm/vllm-openai:v0.11.2   \
  --model /model/huggingface   \  # Points to the base model created by Verl
  --enable-lora   --lora-modules my-lora=/model/lora_adapter
```

but this works

```
docker run --gpus all   -v ~/.cache/huggingface:/root/.cache/huggingface   \
  -v /my-local/base/models:/models   \
  -v /my-model/global_step_4/actor/lora_adapter:/lora_adapter   \
  --env "HF_TOKEN=$HF_TOKEN"   -p 8000:8000   --ipc=host   vllm/vllm-openai:v0.11.2   \
  --model /models/Qwen/Qwen3-0.6B   \ # Points to the base model from HF hub
  --enable-lora   --lora-modules my-lora=/lora_adapter
```

**Comment by thvasilo** [comment]
Can confirm that after applying @yurekami 's fix by doing 

```
# I'm already in a new ven
gh pr checkout 4770
pip install -e .
```

I was able to train a new model and then directly load the model  produced under the `huggingface/` path.

Question for @yurekami though, is the adapter output now duplicated?

Here's what I see

```
.venv_local_verl) ubuntu@ip-10-4-196-159:~/outputs/llm-recipe-lora-20260128_014502/global_step_4/actor$ tree -h
[ 32K]  .
├── [  46]  fsdp_config.json
├── [ 32K]  huggingface
│   ├── [ 707]  added_tokens.json
│   ├── [4.1K]  chat_template.jinja
│   ├── [1.4K]  config.json
│   ├── [ 214]  generation_config.json
│   ├── [ 32K]  lora_adapter
│   │   ├── [1.1K]  adapter_config.json
│   │   └── [ 39M]  adapter_model.safetensors
│   ├── [1.6M]  merges.txt
│   ├── [2.8G]  model.safetensors
│   ├── [ 613]  special_tokens_map.json
│   ├── [ 11M]  tokenizer.json
│   ├── [5.3K]  tokenizer_config.json
│   └── [2.6M]  vocab.json
└── [ 32K]  lora_adapter
    ├── [1.2K]  adapter_config.json
    └── [ 39M]  adapter_model.safetensors

3 directories, 15 files
```

**Comment by thvasilo** [comment]
Hi @vermouth1992 can we take another look at this PR?

I see there was a related commit at https://github.com/verl-project/verl/pull/5326 but as far as I can tell, the FSDP checkpoint manager still produces checkpoints with unnormalized layer names?

**Comment by yurekami** [comment]
@thvasilo Answering your question from January first, sorry it sat this long: yes, the adapter was duplicated. `actor/lora_adapter/` came from the pre-existing dump in `fsdp_workers.py::save_checkpoint`, written from the live `peft_config` (correct `lora_alpha` — your 1.2K `adapter_config.json`). My `huggingface/lora_adapter/` was a second copy with a config reconstructed from tensor shapes (`lora_alpha=0` — the 1.1K one). Same weights, but my copy had the worse config; if you kept using `actor/lora_adapter/` you were fine. The duplicate dump is gone in the update.

Your 2026-03-31 read of #5326 is also correct. #5326 only persists `lora_train_meta.json` and teaches the offline `model_merger` to use it for `lora_alpha`; it never touched the checkpoint-time HF save. On current main, `FSDPCheckpointManager.save_checkpoint` still does `get_fsdp_full_state_dict` → `save_pretrained(hf_local_path, state_dict=state_dict)` (`verl/utils/checkpoint/fsdp_checkpoint_manager.py:344,381`) with no PEFT key handling, so `huggingface/` still contains `base_model.model.*` / `*.base_layer.weight` keys plus raw `lora_A/lora_B` tensors. Two things changed since this PR was written: `fsdp_workers.py` was deleted in the engine refactor, so main no longer writes *any* `lora_adapter/` at save time; and main gained `normalize_peft_param_name` (`verl/utils/fsdp_utils.py:842`) doing exactly this PR's key normalization — but it's only wired into rollout weight sync, not checkpointing.

I've rebuilt the branch on current main accordingly: (1) the HF save path now normalizes the PEFT state dict by reusing `normalize_peft_param_name`, and (2) a single `lora_adapter/` is written whose config comes from the real `peft_config` (no reconstruction, no copy under `huggingface/`). CPU tests cover the load round-trip and the config sourcing. Since every original code site died in the refactor, this is effectively a new diff under the old PR number.

@vermouth1992 the defect still reproduces on main (`huggingface/` output from LoRA runs is not loadable by vLLM/HF); PTAL.
